bulk-updater/src/cli.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MPL-2.0
+
+// Copyright (C) 2023  Soc Virnyl Estela
+
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{io, path::PathBuf};
+
+use clap::Parser;
+use tracing;
+
+use crate::operations;
+
+#[derive(Parser, Debug)]
+#[command(
+    author,
+    name = "bulk_updater",
+    about = "OBS Service Cargo Bulk Updater for Rust projects.",
+    after_long_help = "Set verbosity and tracing through `RUST_LOG` environmental variable e.g. `RUST_LOG=trace`
+
+Bugs can be reported on GitHub: https://github.com/openSUSE/obs-service-cargo_vendor/issues",
+    max_term_width = 120
+)]
+pub struct BulkUpdaterOpts {
+    #[arg(
+        long,
+        default_value = "home:firstyear:branches",
+        help = "Run bulk updater in this OBS project."
+    )]
+    basepath: PathBuf,
+    #[arg(
+        long,
+        default_value_t = false,
+        help = "Wheter to YOLO the bulk updates."
+    )]
+    yolo: bool,
+    #[arg(
+        long,
+        default_value_t = false,
+        help = "Whether to submit the packages manually or not."
+    )]
+    findout: bool,
+    #[arg(
+        long,
+        default_value = "Automatic update of vendored dependencies",
+        help = "Insert custom message for submitting updated package to OBS."
+    )]
+    message: String,
+    #[arg(long, help = "List of Rust packages to be updated.")]
+    packages: Vec<PathBuf>,
+    #[arg(
+        long,
+        default_value = "auto",
+        default_missing_value = "always",
+        value_name = "WHEN",
+        help = "Whether WHEN to color output or not"
+    )]
+    pub color: clap::ColorChoice,
+}
+
+impl BulkUpdaterOpts {
+    pub fn run(self) -> io::Result<()> {
+        let mut pkgpaths: Vec<PathBuf> = Vec::new();
+        for pkgname in self.packages.iter() {
+            tracing::info!(
+                "⏫ Checkout or update in progress for {}",
+                pkgname.to_string_lossy()
+            );
+            pkgpaths.push(operations::checkout_or_update(
+                &pkgname.to_string_lossy(),
+                &self.basepath,
+            )?);
+            tracing::info!("🥳 Updated {}", pkgname.to_string_lossy());
+        }
+        for pkgpath in pkgpaths.iter() {
```

Comment:
For bonus marks, make this par_iter with rayon since these can all update in parallel.

Reply:
yeah i could try using rayon. it's smaller than tokio